File: node/src/openai/openai_tools.ts
import OpenAI from "openai";
import dotenv from "dotenv";
import path from "path";

import {
  ChatCompletionChunk,
  ChatCompletionTool,
} from "openai/resources/chat/completions";

import { Stream } from "openai/streaming";

dotenv.config({ path: path.resolve(__dirname, "../../.env") });

const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

const SYS_PROMPT = `
You are a friendly chatbot answering the user's questions.
`;

const tools: ChatCompletionTool[] = [
  {
    type: "function",
    function: {
      name: "getWeather",
      description: "Get weather information about a city",
      parameters: {
        type: "object",
        properties: {
          city: {
            type: "string",
            description: "Name of the city",
          },
        },
        required: ["city"],
      },
    },
  },
];

async function llmCall(
  messages: any[],
  stream = false,
  tools?: ChatCompletionTool[]
) {
  return await openai.chat.completions.create({
    model: "gpt-4o",
    messages: messages,
    temperature: 0.75,
    tools: tools,
    tool_choice: "auto",
    stream: stream,
  });
}

async function askAboutWeather() {
  const user_msg = "How's the weather in Helsinki?";

  const messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [
    { role: "system", content: SYS_PROMPT },
    { role: "user", content: user_msg },
  ];

  const response = (await llmCall(
    messages,
    false,
    tools
  )) as OpenAI.Chat.Completions.ChatCompletion;

  const tool_calls = response.choices[0].message.tool_calls;
  if (tool_calls) {
    for (const tool_call of tool_calls) {
      messages.push({
        role: "assistant",
        tool_calls: [tool_call],
      });

      const tool_name = tool_call.function.name;
      const tool_args = JSON.parse(tool_call.function.arguments);

      console.log(`LLM called tool: ${tool_name} with args:`, tool_args);

      if (tool_name === "getWeather") {
        const temperature = Math.floor(Math.random() * 21);
        const weather_types = ["sunny", "cloudy", "raining"];
        const weather_type =
          weather_types[Math.floor(Math.random() * weather_types.length)];

        const city = tool_args.city;
        messages.push({
          role: "tool",
          tool_call_id: tool_call.id,
          content: `The weather in ${city} today: ${temperature}°, ${weather_type}`,
        });
      } else {
        throw new Error(`Unknown tool ${tool_name}`);
      }
    }

    const stream = (await llmCall(
      messages,
      true,
      tools
    )) as Stream<OpenAI.Chat.Completions.ChatCompletionChunk>;

    for await (const part of stream) {
      process.stdout.write(part.choices[0]?.delta?.content || "");
    }
  }
}

askAboutWeather();

File: python/openai/openai_tools.py
import json
import os
import random
from typing import List
from openai import OpenAI
from openai.types.chat import ChatCompletionToolParam
from openai.types.chat.chat_completion_message_tool_call import (
    ChatCompletionMessageToolCall,
)
from dotenv import load_dotenv

load_dotenv()

llm = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

SYS_PROMPT = """
You are a friendly chatbot answering the user's questions.
"""

tools: List[ChatCompletionToolParam] = [
    {
        "type": "function",
        "function": {
            "name": "getWeather",
            "description": "Get weather information about a city",
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {
                        "type": "string",
                        "description": "Name of the city",
                    }
                },
                "required": ["city"],
            },
        },
    }
]


def handle_tool_call(tool_call: ChatCompletionMessageToolCall):
    tool_name = tool_call.function.name
    tool_args = json.loads(tool_call.function.arguments)

    print(f"LLM called tool: {tool_name} with args: {tool_args}")

    temperature = random.randint(0, 20)
    weather_type = random.choice(["sunny", "cloudy", "raining"])

    if tool_name == "getWeather":
        city = tool_args["city"]
        return {
            "role": "tool",
            "tool_call_id": tool_call.id,
            "name": "getWeather",
            "content": f"The weather in {city} today: {temperature}°, {weather_type}",
        }
    else:
        raise ValueError(f"Unknown tool {tool_name}")


def call_llm(messages, stream=False, tools=None):
    response = llm.chat.completions.create(
        model="gpt-4o",
        max_tokens=1024,
        messages=messages,
        stream=stream,
        tools=tools,
    )

    return response


user_msg = "What's the weather like in Helsinki?"

messages = [
    {"role": "system", "content": SYS_PROMPT},
    {"role": "user", "content": user_msg},
]

response = call_llm(messages, stream=False, tools=tools)
tool_calls = response.choices[0].message.tool_calls

if tool_calls:
    for tool_call in tool_calls:
        messages.append({"role": "assistant", "tool_calls": [tool_call]})
        tool_response = handle_tool_call(tool_call=tool_call)
        messages.append(tool_response)

    stream = call_llm(messages=messages, stream=True)
    for chunk in stream:
        content = chunk.choices[0].delta.content or ""
        print(content, end="", flush=True)
else:
    print(response.choices[0].message.content)